File: sdk-python/orchestrator/spec.py
"""Workflow spec: parsing, validation and template resolution.

Language-neutral template -> typed objects. Mirrors spec/spec.schema.json.
"""
from __future__ import annotations

import json
import re
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional

_DURATION_RE = re.compile(r"^(\d+)(ms|s|m|h)$")
_TEMPLATE_RE = re.compile(r"\$\{([^}]+)\}")
_UNITS = {"ms": 0.001, "s": 1.0, "m": 60.0, "h": 3600.0}


def parse_duration(text: Optional[str], default: Optional[float] = None) -> Optional[float]:
    """'500ms' -> 0.5 seconds. None/absent -> default."""
    if text is None:
        return default
    m = _DURATION_RE.match(str(text).strip())
    if not m:
        raise SpecError(f"invalid duration: {text!r}")
    return int(m.group(1)) * _UNITS[m.group(2)]


class SpecError(ValueError):
    """Raised when a spec is structurally invalid."""


@dataclass
class RateLimit:
    requests: int
    per: float  # seconds
    tokens: Optional[int] = None
    per_tokens: Optional[float] = None

    @staticmethod
    def from_dict(d: Optional[dict]) -> "Optional[RateLimit]":
        if not d:
            return None
        return RateLimit(
            requests=int(d["requests"]),
            per=parse_duration(d["per"]),
            tokens=d.get("tokens"),
            per_tokens=parse_duration(d.get("per_tokens")),
        )


@dataclass
class Retry:
    max: int = 0
    backoff: str = "exponential"
    base: float = 0.5
    jitter: bool = True

    @staticmethod
    def from_dict(d: Optional[dict]) -> "Optional[Retry]":
        if not d:
            return None
        return Retry(
            max=int(d.get("max", 0)),
            backoff=d.get("backoff", "exponential"),
            base=parse_duration(d.get("base"), 0.5),
            jitter=bool(d.get("jitter", True)),
        )


@dataclass
class Provider:
    name: str
    rate_limit: Optional[RateLimit] = None


@dataclass
class Task:
    id: str
    skill: str
    provider: Optional[str] = None
    depends_on: List[str] = field(default_factory=list)
    inputs: Dict[str, Any] = field(default_factory=dict)
    retry: Optional[Retry] = None
    timeout: Optional[float] = None
    on_error: Optional[str] = None
    idempotent: bool = False


@dataclass
class Workflow:
    name: str
    tasks: List[Task]
    version: int = 1
    max_parallel: int = 8
    rate_limit: Optional[RateLimit] = None
    providers: Dict[str, Provider] = field(default_factory=dict)
    default_retry: Optional[Retry] = None
    default_timeout: Optional[float] = None
    default_on_error: str = "fail_fast"
    inputs_schema: Dict[str, dict] = field(default_factory=dict)
    output: Optional[str] = None

    # ---- loaders ----
    @staticmethod
    def from_dict(doc: dict) -> "Workflow":
        wf = doc.get("workflow")
        if not isinstance(wf, dict):
            raise SpecError("missing top-level 'workflow' object")

        conc = wf.get("concurrency", {}) or {}
        defaults = wf.get("defaults", {}) or {}
        providers = {
            name: Provider(name, RateLimit.from_dict((p or {}).get("rate_limit")))
            for name, p in (wf.get("providers", {}) or {}).items()
        }
        tasks = [
            Task(
                id=t["id"],
                skill=t["skill"],
                provider=t.get("provider"),
                depends_on=list(t.get("depends_on", []) or []),
                inputs=dict(t.get("inputs", {}) or {}),
                retry=Retry.from_dict(t.get("retry")),
                timeout=parse_duration(t.get("timeout")),
                on_error=t.get("on_error"),
                idempotent=bool(t.get("idempotent", False)),
            )
            for t in wf.get("tasks", [])
        ]
        model = Workflow(
            name=wf["name"],
            version=int(wf.get("version", 1)),
            tasks=tasks,
            max_parallel=int(conc.get("max_parallel", 8)),
            rate_limit=RateLimit.from_dict(conc.get("rate_limit")),
            providers=providers,
            default_retry=Retry.from_dict(defaults.get("retry")) or Retry(),
            default_timeout=parse_duration(defaults.get("timeout")),
            default_on_error=defaults.get("on_error", "fail_fast"),
            inputs_schema=wf.get("inputs", {}) or {},
            output=wf.get("output"),
        )
        model.validate()
        return model

    @staticmethod
    def from_json(text: str) -> "Workflow":
        return Workflow.from_dict(json.loads(text))

    @staticmethod
    def from_file(path: str) -> "Workflow":
        with open(path, "r") as fh:
            text = fh.read()
        if path.endswith((".yaml", ".yml")):
            try:
                import yaml  # optional dependency
            except ImportError as exc:  # pragma: no cover
                raise SpecError(
                    "PyYAML is required to load .yaml specs; use a .json spec or `pip install pyyaml`"
                ) from exc
            return Workflow.from_dict(yaml.safe_load(text))
        return Workflow.from_json(text)

    # ---- validation ----
    def validate(self) -> "Workflow":
        ids = [t.id for t in self.tasks]
        if len(ids) != len(set(ids)):
            raise SpecError("duplicate task ids")
        idset = set(ids)
        for t in self.tasks:
            for dep in t.depends_on:
                if dep not in idset:
                    raise SpecError(f"task {t.id!r} depends on unknown task {dep!r}")
            if t.provider and t.provider not in self.providers:
                raise SpecError(f"task {t.id!r} references unknown provider {t.provider!r}")
        self._assert_acyclic()
        return self

    def _assert_acyclic(self) -> None:
        graph = {t.id: list(t.depends_on) for t in self.tasks}
        WHITE, GREY, BLACK = 0, 1, 2
        color = {tid: WHITE for tid in graph}

        def visit(node: str) -> None:
            color[node] = GREY
            for dep in graph[node]:
                if color[dep] == GREY:
                    raise SpecError(f"dependency cycle detected at {node!r} -> {dep!r}")
                if color[dep] == WHITE:
                    visit(dep)
            color[node] = BLACK

        for tid in graph:
            if color[tid] == WHITE:
                visit(tid)

    def task_map(self) -> Dict[str, Task]:
        return {t.id: t for t in self.tasks}


def resolve_template(value: Any, scope: Dict[str, Any]) -> Any:
    """Resolve ${...} references against a scope dict.

    Supported paths: input.<field>, <taskId>.output, <taskId>.output.<field...>.
    A string that is exactly one reference returns the referenced value (preserving
    its type); otherwise references are substituted as text within the string.
    """
    if isinstance(value, dict):
        return {k: resolve_template(v, scope) for k, v in value.items()}
    if isinstance(value, list):
        return [resolve_template(v, scope) for v in value]
    if not isinstance(value, str):
        return value

    full = _TEMPLATE_RE.fullmatch(value.strip())
    if full:
        return _lookup(full.group(1).strip(), scope)

    def repl(m: "re.Match[str]") -> str:
        v = _lookup(m.group(1).strip(), scope)
        return v if isinstance(v, str) else json.dumps(v)

    return _TEMPLATE_RE.sub(repl, value)


def _lookup(path: str, scope: Dict[str, Any]) -> Any:
    parts = path.split(".")
    node: Any = scope
    for p in parts:
        if isinstance(node, dict) and p in node:
            node = node[p]
        else:
            raise SpecError(f"unresolved template reference: ${{{path}}}")
    return node

File: engine/internal/spec/spec.go
// Package spec parses and validates the language-neutral workflow template.
// Mirrors spec/spec.schema.json; the canonical source of the data model.
package spec

import (
	"encoding/json"
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

// ---------------------------------------------------------------------------
// Duration
// ---------------------------------------------------------------------------

// Duration is seconds as a float64. Parsed from strings like "500ms", "120s", "5m", "1h".
type Duration float64

var durationRE = regexp.MustCompile(`^(\d+)(ms|s|m|h)$`)

func ParseDuration(s string) (Duration, error) {
	m := durationRE.FindStringSubmatch(strings.TrimSpace(s))
	if m == nil {
		return 0, fmt.Errorf("invalid duration %q (expected e.g. 500ms, 30s, 5m, 1h)", s)
	}
	n, _ := strconv.ParseFloat(m[1], 64)
	switch m[2] {
	case "ms":
		return Duration(n / 1000), nil
	case "s":
		return Duration(n), nil
	case "m":
		return Duration(n * 60), nil
	default: // h
		return Duration(n * 3600), nil
	}
}

func (d *Duration) UnmarshalJSON(b []byte) error {
	var s string
	if err := json.Unmarshal(b, &s); err != nil {
		return err
	}
	v, err := ParseDuration(s)
	if err != nil {
		return err
	}
	*d = v
	return nil
}

// ---------------------------------------------------------------------------
// Sub-types
// ---------------------------------------------------------------------------

type RateLimit struct {
	Requests  int       `json:"requests"`
	Per       Duration  `json:"per"`
	Tokens    int       `json:"tokens,omitempty"`
	PerTokens *Duration `json:"per_tokens,omitempty"`
}

type Retry struct {
	Max     int      `json:"max"`
	Backoff string   `json:"backoff"` // none | fixed | exponential
	Base    Duration `json:"base"`
	Jitter  bool     `json:"jitter"`
}

func DefaultRetry() Retry {
	return Retry{Max: 0, Backoff: "exponential", Base: Duration(0.5), Jitter: true}
}

type Provider struct {
	Name      string     `json:"-"`
	RateLimit *RateLimit `json:"rate_limit,omitempty"`
}

type InputSpec struct {
	Type     string `json:"type"`
	Required bool   `json:"required"`
	Default  any    `json:"default,omitempty"`
}

// ---------------------------------------------------------------------------
// Task
// ---------------------------------------------------------------------------

type Task struct {
	ID         string         `json:"id"`
	Skill      string         `json:"skill"`
	Provider   string         `json:"provider,omitempty"`
	DependsOn  []string       `json:"depends_on,omitempty"`
	Inputs     map[string]any `json:"inputs,omitempty"`
	Retry      *Retry         `json:"retry,omitempty"`
	Timeout    *Duration      `json:"timeout,omitempty"`
	OnError    string         `json:"on_error,omitempty"` // fail_fast | continue
	Idempotent bool           `json:"idempotent,omitempty"`
}

// EffectiveRetry returns the task's retry policy, falling back to the workflow default.
func (t *Task) EffectiveRetry(dflt Retry) Retry {
	if t.Retry != nil {
		return *t.Retry
	}
	return dflt
}

// EffectiveTimeout returns the task's timeout, falling back to the workflow default.
func (t *Task) EffectiveTimeout(dflt *Duration) *Duration {
	if t.Timeout != nil {
		return t.Timeout
	}
	return dflt
}

// EffectiveOnError returns the task's error policy, falling back to the workflow default.
func (t *Task) EffectiveOnError(dflt string) string {
	if t.OnError != "" {
		return t.OnError
	}
	return dflt
}

// ---------------------------------------------------------------------------
// Workflow
// ---------------------------------------------------------------------------

type Workflow struct {
	Name           string               `json:"name"`
	Version        int                  `json:"version,omitempty"`
	MaxParallel    int                  `json:"-"` // from concurrency.max_parallel
	RateLimit      *RateLimit           `json:"-"` // from concurrency.rate_limit
	Providers      map[string]*Provider `json:"providers,omitempty"`
	DefaultRetry   Retry                `json:"-"`
	DefaultTimeout *Duration            `json:"-"`
	DefaultOnError string               `json:"-"`
	InputsSchema   map[string]InputSpec `json:"inputs,omitempty"`
	Tasks          []*Task              `json:"tasks"`
	Output         string               `json:"output,omitempty"`
}

// TaskMap returns tasks indexed by ID.
func (w *Workflow) TaskMap() map[string]*Task {
	m := make(map[string]*Task, len(w.Tasks))
	for _, t := range w.Tasks {
		m[t.ID] = t
	}
	return m
}

// ---------------------------------------------------------------------------
// Parsing
// ---------------------------------------------------------------------------

// raw mirrors the on-disk structure for JSON unmarshalling.
type raw struct {
	Workflow struct {
		Name        string `json:"name"`
		Version     int    `json:"version"`
		Concurrency *struct {
			MaxParallel int        `json:"max_parallel"`
			RateLimit   *RateLimit `json:"rate_limit"`
		} `json:"concurrency"`
		Providers map[string]*Provider `json:"providers"`
		Defaults  *struct {
			Retry   *Retry    `json:"retry"`
			Timeout *Duration `json:"timeout"`
			OnError string    `json:"on_error"`
		} `json:"defaults"`
		Inputs map[string]InputSpec `json:"inputs"`
		Tasks  []*Task              `json:"tasks"`
		Output string               `json:"output"`
	} `json:"workflow"`
}

// Parse decodes and validates a workflow spec from JSON bytes.
func Parse(data []byte) (*Workflow, error) {
	var r raw
	if err := json.Unmarshal(data, &r); err != nil {
		return nil, fmt.Errorf("spec parse error: %w", err)
	}
	rw := r.Workflow
	if rw.Name == "" {
		return nil, fmt.Errorf("spec: workflow.name is required")
	}
	if len(rw.Tasks) == 0 {
		return nil, fmt.Errorf("spec: workflow.tasks must have at least one task")
	}

	wf := &Workflow{
		Name:           rw.Name,
		Version:        rw.Version,
		MaxParallel:    8,
		Providers:      rw.Providers,
		DefaultRetry:   DefaultRetry(),
		DefaultOnError: "fail_fast",
		InputsSchema:   rw.Inputs,
		Tasks:          rw.Tasks,
		Output:         rw.Output,
	}
	if wf.Version == 0 {
		wf.Version = 1
	}
	if wf.Providers == nil {
		wf.Providers = map[string]*Provider{}
	}
	for name, p := range wf.Providers {
		p.Name = name
	}
	if rw.Concurrency != nil {
		if rw.Concurrency.MaxParallel > 0 {
			wf.MaxParallel = rw.Concurrency.MaxParallel
		}
		wf.RateLimit = rw.Concurrency.RateLimit
	}
	if rw.Defaults != nil {
		if rw.Defaults.Retry != nil {
			wf.DefaultRetry = *rw.Defaults.Retry
		}
		wf.DefaultTimeout = rw.Defaults.Timeout
		if rw.Defaults.OnError != "" {
			wf.DefaultOnError = rw.Defaults.OnError
		}
	}
	if err := wf.Validate(); err != nil {
		return nil, err
	}
	return wf, nil
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

// Validate checks for duplicate IDs, unknown deps/providers, and cycles.
func (w *Workflow) Validate() error {
	ids := make(map[string]bool, len(w.Tasks))
	for _, t := range w.Tasks {
		if ids[t.ID] {
			return fmt.Errorf("spec: duplicate task id %q", t.ID)
		}
		ids[t.ID] = true
	}
	for _, t := range w.Tasks {
		for _, dep := range t.DependsOn {
			if !ids[dep] {
				return fmt.Errorf("spec: task %q depends on unknown task %q", t.ID, dep)
			}
		}
		if t.Provider != "" {
			if _, ok := w.Providers[t.Provider]; !ok {
				return fmt.Errorf("spec: task %q references unknown provider %q", t.ID, t.Provider)
			}
		}
	}
	return w.detectCycle()
}

func (w *Workflow) detectCycle() error {
	const (
		white = iota
		grey
		black
	)
	graph := make(map[string][]string, len(w.Tasks))
	for _, t := range w.Tasks {
		graph[t.ID] = t.DependsOn
	}
	color := make(map[string]int, len(w.Tasks))
	var visit func(node string) error
	visit = func(node string) error {
		color[node] = grey
		for _, dep := range graph[node] {
			if color[dep] == grey {
				return fmt.Errorf("spec: dependency cycle detected at %q → %q", node, dep)
			}
			if color[dep] == white {
				if err := visit(dep); err != nil {
					return err
				}
			}
		}
		color[node] = black
		return nil
	}
	for id := range graph {
		if color[id] == white {
			if err := visit(id); err != nil {
				return err
			}
		}
	}
	return nil
}
